src/Components/ScrivitoExtensions/ChatbotTab/useChatCompletion.js

```diff
@@ -0,0 +1,117 @@
+
+export function useChatCompletion({ apiKey, instanceId, model, user }) {
+  const [messages, setMessages] = useState([]);
+  const [loading, setLoading] = useState(false);
+  const [completionMessage, setCompletionMessage] = useState(null);
+
+  const messagesWithCompletion = useMemo(
+    () => (completionMessage ? messages.concat(completionMessage) : messages),
+    [messages, completionMessage]
+  );
+
+  return {
+    loading,
+    messages: messagesWithCompletion,
+    submitPrompt: (prompt) => {
+      setCompletionMessage(null);
+      setLoading(true);
+      // @ts-ignore
+      setMessages((m) => {
+        const messagesWithPrompt = [...m, ...prompt];
+        startStreaming({
+          apiKey,
+          instanceId,
+          model,
+          messages: messagesWithPrompt,
+          setCompletionMessage,
+          setLoading,
+          setMessages,
+          user,
+        });
+        return messagesWithPrompt;
+      });
+    },
+    abortResponse: () => {},
+    resetMessages: () => {
+      setCompletionMessage(null);
+      setMessages([]);
+    },
+    setMessages,
+  };
+}
+
+let OPENAI_API_KEY =
+  // @ts-ignore
+  typeof import.meta.env === "undefined"
+    ? // @ts-ignore
+      process.env.OPENAI_API_KEY
+    : // @ts-ignore
+      import.meta.env.OPENAI_API_KEY;
+
+async function startStreaming({
+  apiKey,
+  instanceId,
+  messages,
+  model,
+  setCompletionMessage,
+  setLoading,
+  setMessages,
+  user,
+}) {
+  const client = new OpenAI({
+    apiKey: OPENAI_API_KEY || apiKey,
+    baseURL: OPENAI_API_KEY
+      ? "https://api.openai.com/v1"
+      : "https://i7ukqy3mhy3nzkn3dutmmzdx440xgtjk.lambda-url.eu-west-1.on.aws?ignore=",
+
+    defaultQuery: { tenant_id: instanceId },
+    defaultHeaders: { Accept: "*/*" },
+    dangerouslyAllowBrowser: true,
+    fetch: async (url, init) => {
+      return fetch(url, {
+        ...init,
+        headers: cleanHeaders(init?.headers),
+        body: JSON.stringify(JSON.parse(init?.body)),
+      });
+    },
+  });
+
+  const stream = await client.beta.chat.completions.stream({
+    model,
+    messages,
+    stream: true,
+    user,
+  });
+
+  stream.on("error", () => {
+    setLoading(false);
+  });
+
+  stream.on("content", () => {
+    const message = stream.currentChatCompletionSnapshot?.choices[0].message;
+    if (message) setCompletionMessage({ ...message });
+  });
+
+  return stream.finalChatCompletion().then(({ choices }) => {
+    setCompletionMessage(null);
+    const { tool_calls, ...otherMessageAttributes } = choices[0].message;
+    setMessages(messages.concat({ ...otherMessageAttributes }));
+    setLoading(false);
+  });
+}
+
+function cleanHeaders(headers = {}) {
+  return Object.fromEntries(
+    Object.entries(headers)
+      .filter(([k]) => !k.startsWith("x-"))
+      .map(([k, v]) => [
+        k
+          .replace("content-type", "Content-Type")
+          .replace("authorization", "Authorization")
+          .replace("accept", "Accept"),
```

Comment:
FMI Why do we need this? Aren't HTTP header case-insensitive?

Reply:
@dcsaszar can you share why this was added in your branch?

Reply:
It works for me without this header camelcaseing, so I assume it's just cruft?

Reply:
Removing the "x-"-headers is currently still necessary, though. 

Apparently OpenAI does a lot of secret tracking: 
```
x-stainless-lang js
x-stainless-package-version 4.56.0
x-stainless-os Unknown
x-stainless-arch unknown
x-stainless-runtime browser:chrome
x-stainless-runtime-version 127.0.0
x-stainless-helper-method stream
```

And this leads to that:
```
blocked by CORS policy: 
Request header field x-stainless-os is not allowed 
by Access-Control-Allow-Headers in preflight response.
```

IMHO we should just accept these headers in our backend's CORS policy, because I don't like this need to "patch" the OpenAI library.

The reason is that it's not sustainable: If this approach succeeds, may developers will do this. And I don't feel like maintaining documentation about this ("This is how you patch the OpenAI client so that it can talk to our backend...").

It's much simpler to just fix it on the backend, instead, IMHO.

Reply:
And who knows, maybe that analytics info can also help us debug at some point, e.g. knowing that the openai library was used and in which version can help us understand problems in the service.

Reply:
> maybe that analytics info can also help us debug at some point

Currently we don't do any logging of request data in the proxy. If you want this, please let me know 🙂

Reply:
> It's much simpler to just fix it on the backend, instead, IMHO.

At the time I added the filter client-side, we still used the AWS Lambda config for CORS and it would have been a pain to add all these headers manually.
I hope this is easier with the `cors` module we now use in the proxy.

Reply:
And the case conversion can probably be removed. I think I added this while investigating why the proxy was happy with the previous lib but not the openAI lib. I wanted the network requests look identical.

Reply:
> I hope this is easier with the cors module we now use in the proxy.

Yep:

> If not specified, defaults to reflecting the headers specified in the request’s [[1](https://expressjs.com/en/resources/middleware/cors.html#configuration-options)]

🥳